### PR TITLE
Added samples for java/kotlin

### PR DESCRIPTION
**Pull request by BavovnaIsTaken**

### CODE REVIEW

File: samples/android-kotlin-coroutines/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  - Android Studio 3.2.1+
+  - [Sample Backend for Java](https://github.com/VirgilSecurity/sample-backend-java) running on localhost:3000.
+  
+### How to start
+Don't forget to setup [Sample Backend for Java](https://github.com/VirgilSecurity/sample-backend-java) first. It is a mandatory part of this sample. After this do next few steps:
+
+```
+  - git clone https://github.com/VirgilSecurity/virgil-e3kit-kotlin
+  - cd virgil-e3kit-kotlin
+  - git checkout dev
```
Comment (andrii-iakovenko) [inline samples/android-kotlin-coroutines/README.md]: I think this is not necessary.

File: samples/android-java-firebase-function/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  - Android Studio 3.2.1+
+  - [Firebase function](https://github.com/VirgilSecurity/virgil-e3kit-firebase-func) deployed successfully.
+
+### How to start
+Don't forget to deploy [Firebase function](https://github.com/VirgilSecurity/virgil-e3kit-firebase-func) first. It is a mandatory part of this sample. After this do next few steps:
+
+```
+  - git clone https://github.com/VirgilSecurity/virgil-e3kit-kotlin
+  - cd virgil-e3kit-kotlin
+  - git checkout dev
```
Comment (andrii-iakovenko) [inline samples/android-java-firebase-function/README.md]: Master branch also could be used

File: samples/android-java-firebase-function/app/build.gradle [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    implementation 'com.android.support:animated-vector-drawable:28.0.0'
+    implementation 'com.android.support:support-media-compat:28.0.0'
+    implementation 'com.android.support.constraint:constraint-layout:1.1.3'
+
+    // Firebase
+    implementation 'com.google.firebase:firebase-core:16.0.6'
+    implementation 'com.google.firebase:firebase-auth:16.1.0'
+
+    // E3Kit
+    implementation "com.virgilsecurity:ethree-kotlin:0.3.3"
```
Comment (andrii-iakovenko) [inline samples/android-java-firebase-function/app/build.gradle]: 0.3.4?

File: samples/android-java-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplejavafirebase/EThreeActivity.java [hunk 4]
```diff
@@ -0,0 +1,377 @@
+/*
+ * Copyright (c) 2015-2018, Virgil Security, Inc.
```
Comment (andrii-iakovenko) [inline samples/android-java-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplejavafirebase/EThreeActivity.java]: 2015-2019

File: samples/android-java-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplejavafirebase/EThreeActivity.java [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+                    runOnUiThread(new Runnable() {
+                        @Override public void run() {
+                            tvText.setText("Success. Sample finished it's work.\n\n" + encryptedText);
+                        }
+                    });
+
+                    // encrypts provided text and returns encrypted byte array.
+                    byte[] encryptedData = eThreeUserTwo.encrypt(data, new ArrayList<>(result.values()));
+
+                    System.out.println("encryptedText: \n" + encryptedText);
```
Comment (andrii-iakovenko) [inline samples/android-java-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplejavafirebase/EThreeActivity.java]: Use Log.d instead

File: samples/android-java-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplejavafirebase/EThreeActivity.java [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+                    // You can convert byte[] to Base64 String to easily transfer it to the server, or to print, etc.
+                    System.out.println("encryptedData: \n" + ConvertionUtils.toBase64String(encryptedData));
```
Comment (andrii-iakovenko) [inline samples/android-java-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplejavafirebase/EThreeActivity.java]: Use Log.d instead

File: samples/android-java-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplejavafirebase/EThreeActivity.java [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+    /**
+     * This method exchanges provided authToken for a Virgil JWT.
+     *
+     * @param authToken from your authentication system that signals that user is authenticated successfully.
+     *
+     * @return Virgil JWT base64 string representation.
+     */
+    private String getVirgilJwt(String authToken) {
+        try {
+            String url = "<your-firebase-function-url>/virgil-jwt";
```
Comment (andrii-iakovenko) [inline samples/android-java-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplejavafirebase/EThreeActivity.java:339]: Why not to move this to BuildConfig?

File: samples/android-java-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplejavafirebase/EThreeActivity.java [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+                }
+                br.close();
+                JSONObject jsonObject = new JSONObject(sb.toString());
+
+                return jsonObject.getString("token");
+            } else {
+                throw new RuntimeException("Some connection error");
+            }
+        } catch (IOException exception) {
+            exception.printStackTrace();
```
Comment (andrii-iakovenko) [inline samples/android-java-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplejavafirebase/EThreeActivity.java]: Log.e

File: samples/android-java-firebase-function/app/src/main/res/layout/activity_ethree.xml [hunk 9]
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="utf-8"?><!--
+  ~ Copyright (c) 2015-2018, Virgil Security, Inc.
```
Comment (andrii-iakovenko) [inline samples/android-java-firebase-function/app/src/main/res/layout/activity_ethree.xml]: 2015-2019

File: samples/android-kotlin-coroutines/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlincoroutines/EThreeActivity.kt [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+            // encrypt method encrypts provided text and converts it to Base64 String format.
+            val encryptedText = eThreeUserTwo.encrypt(text, keysResult.values.toList())
+
+            runOnUiThread { tvText.text = "Success. Sample finished it's work.\n" +
+                                          "\n" + encryptedText }
+
+            // encrypts provided text and returns encrypted byte array.
+            val encryptedData = eThreeUserTwo.encrypt(data, keysResult.values.toList())
+
+            println("encryptedText: \n$encryptedText")
```
Comment (andrii-iakovenko) [inline samples/android-kotlin-coroutines/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlincoroutines/EThreeActivity.kt]: Log?

File: samples/android-kotlin-coroutines/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlincoroutines/EThreeActivity.kt [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+     * In this function you have to use your authentication - Firebase, Google, Facebook, or any other auth (as well
+     * as some kind of your custom authentication service).
+     *
+     * Provided [identity] is used for user authentication. You will receive authentication token that you can later
+     * exchange for a Virgil JWT. Or error if something went wrong.
+     */
+    private fun authenticateAsync(
+        identity: String
+    ): Deferred<String> = GlobalScope.async {
+        val baseUrl = "http://10.0.2.2:3000/authenticate"
```
Comment (andrii-iakovenko) [inline samples/android-kotlin-coroutines/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlincoroutines/EThreeActivity.kt:129]: Could we move it to BuildConfig?

File: samples/android-kotlin-firebase-function/app/build.gradle [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    implementation 'com.android.support:animated-vector-drawable:28.0.0'
+    implementation 'com.android.support:support-media-compat:28.0.0'
+    implementation 'com.android.support.constraint:constraint-layout:1.1.3'
+
+    // Firebase
+    implementation 'com.google.firebase:firebase-core:16.0.6'
+    implementation 'com.google.firebase:firebase-auth:16.1.0'
+
+    // E3Kit
+    implementation "com.virgilsecurity:ethree-kotlin:0.3.3"
```
Comment (andrii-iakovenko) [inline samples/android-kotlin-firebase-function/app/build.gradle]: 0.3.4

File: samples/android-kotlin-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlinfirebase/EThreeActivity.kt [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+            // encrypt method encrypts provided text and converts it to Base64 String format.
+            val encryptedText = eThreeUserTwo.encrypt(text, ArrayList(result.values))
+
+            runOnUiThread { tvText.text = "Success. Sample finished it's work.\n" +
+                                          "\n" + encryptedText }
+
+            // encrypts provided text and returns encrypted byte array.
+            val encryptedData = eThreeUserTwo.encrypt(data, ArrayList(result.values))
+
+            println("encryptedText: \n$encryptedText")
```
Comment (andrii-iakovenko) [inline samples/android-kotlin-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlinfirebase/EThreeActivity.kt]: Log

File: samples/android-kotlin-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlinfirebase/EThreeActivity.kt [hunk 14]
```diff
@@ -0,0 +1,299 @@
...
+                // OnResultListener<EThree> will give you initialized instance of EThree SDK in onSuccess method.
+                EThree.initialize(
+                    this@EThreeActivity,
+                    onGetTokenUserOneCallback,
+                    onInitUserOneListener
+                )
+            }
+
+            override fun onError(throwable: Throwable?) {
+                throwable!!.printStackTrace()
```
Comment (andrii-iakovenko) [inline samples/android-kotlin-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlinfirebase/EThreeActivity.kt]: Use Log instead

File: samples/android-kotlin-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlinfirebase/EThreeActivity.kt [hunk 15]
```diff
@@ -0,0 +1,299 @@
...
+
+                EThree.initialize(
+                    this@EThreeActivity,
+                    onGetTokenUserTwoCallback,
+                    onInitUserTwoListener
+                )
+            }
+
+            override fun onError(throwable: Throwable?) {
+                throwable!!.printStackTrace()
```
Comment (andrii-iakovenko) [inline samples/android-kotlin-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlinfirebase/EThreeActivity.kt]: Use Log instead

File: samples/android-kotlin-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlinfirebase/EThreeActivity.kt [hunk 16]
```diff
@@ -0,0 +1,299 @@
...
+                }
+
+                val jsonObject = JSONObject(response)
+
+                return jsonObject.getString("token")
+            } else {
+                throw RuntimeException("Some connection error")
+            }
+        } catch (exception: IOException) {
+            exception.printStackTrace()
```
Comment (andrii-iakovenko) [inline samples/android-kotlin-firebase-function/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlinfirebase/EThreeActivity.kt]: Use logger instead

File: samples/android-kotlin-firebase-function/app/src/main/res/layout/activity_ethree.xml [hunk 17]
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="utf-8"?><!--
+  ~ Copyright (c) 2015-2018, Virgil Security, Inc.
```
Comment (andrii-iakovenko) [inline samples/android-kotlin-firebase-function/app/src/main/res/layout/activity_ethree.xml]: 2015-2019

File: samples/android-kotlin/README.md [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+  - Android Studio 3.2.1+
+  - [Sample Backend for Java](https://github.com/VirgilSecurity/sample-backend-java) running on localhost:3000.
+  
+### How to start
+Don't forget to setup [Sample Backend for Java](https://github.com/VirgilSecurity/sample-backend-java) first. It is a mandatory part of this sample. After this do next few steps:
+
+```
+  - git clone https://github.com/VirgilSecurity/virgil-e3kit-kotlin
+  - cd virgil-e3kit-kotlin
+  - git checkout dev
```
Comment (andrii-iakovenko) [inline samples/android-kotlin/README.md]: Why not master?

File: samples/android-kotlin/app/build.gradle [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+        }
+    }
+}
+
+dependencies {
+    implementation fileTree(dir: 'libs', include: ['*.jar'])
+    implementation"org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+    implementation 'com.android.support:appcompat-v7:28.0.0'
+
+    implementation "com.virgilsecurity:ethree-kotlin:0.3.3"
```
Comment (andrii-iakovenko) [inline samples/android-kotlin/app/build.gradle]: 0.3.4

File: samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt [hunk 20]
```diff
@@ -0,0 +1,310 @@
+package com.android.virgilsecurity.ethreesamplekotlin
```
Comment (andrii-iakovenko) [inline samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt:1]: No file header here

File: samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt [hunk 21]
```diff
@@ -0,0 +1,310 @@
...
+            // encrypt method encrypts provided text and converts it to Base64 String format.
+            val encryptedText = eThreeUserTwo!!.encrypt(text, ArrayList(result.values))
+
+            this@EThreeActivity.runOnUiThread { tvText.text = "Success. Sample finished it's work.\n" +
+                                                              "\n" + encryptedText }
+
+            // encrypts provided text and returns encrypted byte array.
+            val encryptedData = eThreeUserTwo!!.encrypt(data, ArrayList(result.values))
+
+            println("encryptedText: \n$encryptedText")
```
Comment (andrii-iakovenko) [inline samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt]: Use logger instead

File: samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt [hunk 22]
```diff
@@ -0,0 +1,310 @@
...
+     */
+    private fun initUserOne() {
+        // You start your user authentication/authorization (signUp/signIn) here.
+        authenticate(identityOne, object : OnResultListener<String> {
+            override fun onSuccess(value: String) {
+                try {
+                    val `object` = JSONObject(value)
+                    authTokenUserOne = `object`.get("authToken") as String
+                } catch (e: JSONException) {
+                    e.printStackTrace()
```
Comment (andrii-iakovenko) [inline samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt]: Use logger instead

File: samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt [hunk 23]
```diff
@@ -0,0 +1,310 @@
...
+                // OnResultListener<EThree> will give you initialized instance of EThree SDK in onSuccess method.
+                EThree.initialize(
+                    this@EThreeActivity,
+                    onGetTokenUserOneCallback,
+                    onInitUserOneListener
+                )
+            }
+
+            override fun onError(throwable: Throwable) {
+                throwable.printStackTrace()
```
Comment (andrii-iakovenko) [inline samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt]: Use logger instead

File: samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt [hunk 24]
```diff
@@ -0,0 +1,310 @@
...
+     */
+    fun initUserTwo() {
+        // You start your user authentication/authorization (signUp/signIn) here.
+        authenticate(identityTwo, object : OnResultListener<String> {
+            override fun onSuccess(value: String) {
+                try {
+                    val `object` = JSONObject(value)
+                    authTokenUserTwo = `object`.get("authToken") as String
+                } catch (e: JSONException) {
+                    e.printStackTrace()
```
Comment (andrii-iakovenko) [inline samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt]: Use logger instead

File: samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt [hunk 25]
```diff
@@ -0,0 +1,310 @@
...
+    /**
+     * If you use some normal architecture like MVP, MVC, MVVM - you can get rid of the Callback hell. As well
+     * you can wrap all callbacks to RxJava and make it really concise. But you can use just callbacks. It's up to you.
+     */
+    fun initUserTwo() {
+        // You start your user authentication/authorization (signUp/signIn) here.
+        authenticate(identityTwo, object : OnResultListener<String> {
+            override fun onSuccess(value: String) {
+                try {
+                    val `object` = JSONObject(value)
```
Comment (andrii-iakovenko) [inline samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt:183]: Why are you using so strange name?

File: samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt [hunk 26]
```diff
@@ -0,0 +1,310 @@
...
+
+                EThree.initialize(
+                    this@EThreeActivity,
+                    onGetTokenUserTwoCallback,
+                    onInitUserTwoListener
+                )
+            }
+
+            override fun onError(throwable: Throwable) {
+                throwable.printStackTrace()
```
Comment (andrii-iakovenko) [inline samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt]: Use logger instead

File: samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt [hunk 27]
```diff
@@ -0,0 +1,310 @@
...
+     * @param identity         is identityOne of user for authentication
+     * @param onResultListener is a callback where you you will receive authentication token that you can later exchange
+     * for a Virgil JWT. Or error if something went wrong.
+     */
+    private fun authenticate(
+        identity: String,
+        onResultListener: OnResultListener<String>
+    ) {
+        try {
+            val baseUrl = "http://10.0.2.2:3000/authenticate"
```
Comment (andrii-iakovenko) [inline samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt:219]: Let's move it to BuildConfig

File: samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt [hunk 28]
```diff
@@ -0,0 +1,310 @@
...
+            if (httpResult == HttpURLConnection.HTTP_OK) {
+                val response = InputStreamReader(urlConnection.inputStream, "UTF-8").buffered().use {
+                    it.readText()
+                }
+                onResultListener.onSuccess(response)
+            } else {
+                onResultListener.onError(Throwable("Some connection error"))
+            }
+        } catch (exception: JSONException) {
+            exception.printStackTrace()
```
Comment (andrii-iakovenko) [inline samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt]: Use logger instead

File: samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt [hunk 29]
```diff
@@ -0,0 +1,310 @@
...
+            onResultListener.onError(exception)
+        } catch (exception: IOException) {
+            exception.printStackTrace()
```
Comment (andrii-iakovenko) [inline samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt]: Use logger instead

File: samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt [hunk 30]
```diff
@@ -0,0 +1,310 @@
...
+    /**
+     * This method exchanges provided authToken for a Virgil JWT.
+     *
+     * @param authToken from your authentication system that signals that user is authenticated successfully.
+     *
+     * @return Virgil JWT base64 string representation.
+     */
+    private fun getVirgilJwt(authToken: String?): String {
+        try {
+            val baseUrl = "http://10.0.2.2:3000/virgil-jwt"
```
Comment (andrii-iakovenko) [inline samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt:263]: Let's use BuildConfig

File: samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt [hunk 31]
```diff
@@ -0,0 +1,310 @@
...
+                    it.readText()
+                }
+                val jsonObject = JSONObject(response)
+
+                return jsonObject.getString("virgilToken")
+            } else {
+                throw RuntimeException("Some connection error")
+            }
+        } catch (exception: IOException) {
+            exception.printStackTrace()
```
Comment (andrii-iakovenko) [inline samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt]: Use logger instead

File: samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt (hunk 27 above)
Comment (BavovnaIsTaken) [reply samples/android-kotlin/app/src/main/java/com/android/virgilsecurity/ethreesamplekotlin/EThreeActivity.kt:219]: This demo should be stupidly simple, so I think it's better to leave all info in this class

**Review by andrii-iakovenko — APPROVED** [review]

**Pull request merged**